**[RQ4] What is the most common way of payments? Discover the way payments are executed in each borough and visualize the number of payments for any possible means. Then run the Chi-squared test to see whether the method of payment is correlated to the borough. Then, comment the results.**

In [1]:
###month of FEBRUARY

#We imported only the columns useful to answer our Rquestion.
#NB:Payment_type represented by a numeric code signifying how the passenger paid for the trip.  
#1= Credit card; 2= Cash; 3= No charge; 4= Dispute; 5= Unknown; 6= Voided trip.

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import math

data=pd.read_csv('/Users/Enzopc/Desktop/yellow_tripdata_2018-02.csv', usecols=['passenger_count','PULocationID', 'DOLocationID','payment_type','total_amount'])
zone=pd.read_csv('/Users/Enzopc/Desktop/taxi _zone_lookup.csv', usecols=['LocationID', 'Borough'])
datazone=pd.merge(data, zone, how='left', left_on=['PULocationID'],right_on=['LocationID'])

print("Number of rows:", datazone.shape[0])
print("Number of columns: ", datazone.shape[1])
datazone.head()

Number of rows: 8492076
Number of columns:  7


passenger_count  PULocationID  DOLocationID  payment_type  total_amount  \
0                1           145           145             2           4.8   
1                1           145           145             2           3.8   
2                1           236           119             2          19.8   
3                1            82            82             2           4.8   
4                2            82             7             1          15.3   

   LocationID    Borough  
0         145     Queens  
1         145     Queens  
2         236  Manhattan  
3          82     Queens  
4          82     Queens

In [2]:
# What is the most common way of payments? #
datazone['payment_type'].value_counts()

1    5990799
2    2446989
3      42675
4      11613
Name: payment_type, dtype: int64

As we can see from an initial count, it seems that the most common way of payments is by credit card


1: 5990799   2: 2446989    3: 42675    4: 11613

In [3]:
#But now let's see more deeply: 
#Discover the way payments are executed in each borough and visualize the number of payments for any possible means.

contingency_table = pd.crosstab(datazone['payment_type'], datazone['Borough'])
contingency_table

Borough       Bronx  Brooklyn  EWR  Manhattan  Queens  Staten Island  Unknown
payment_type                                                                 
1              2691     61103  397    5491193  339714             50    95651
2              3595     32518   69    2232650  138278             52    39827
3               228      1211   26      34304    5326              5     1575
4                52       269    5       9734    1203              1      349

**Chi-squared test to see whether the method of payment is correlated to the borough.**

H0 (Null Hypothesis): There is no relationship between variable one and variable two.

H1 (Alternative Hypothesis): There is a relationship between variable 1 and variable 2.

If the p-value is significant, you can reject the null hypothesis and claim that the findings support the alternative hypothesis.


Chi-square Assumptions that need to be met in order for the results of the Chi-square test to be trusted:

 -When testing the data, the cells should be frequencies or counts of cases and not percentages. It is okay to convert to percentages after testing the data
 
 -The levels (categories) of the variables being tested are mutually exclusive
 
 -Each participant contributes to only one cell within the Chi-square table
 
 -The groups being tested must be independent
 
 -The value of expected cells should be greater than 5

If all of these assumptions are met, then Chi-square is the correct test to use.

In [4]:
from scipy import stats

stats.chi2_contingency(contingency_table)

(12698.78107337389,
 0.0,
 18,
 array([[4.63203417e+03, 6.70897170e+04, 3.50612395e+02, 5.47991018e+06,
         3.41808990e+05, 7.61894138e+01, 9.69312762e+04],
        [1.89199081e+03, 2.74033229e+04, 1.43210392e+02, 2.23831244e+06,
         1.39614572e+05, 3.11201657e+01, 3.95923426e+04],
        [3.29959423e+01, 4.77908485e+02, 2.49756067e+00, 3.90357224e+04,
         2.43485029e+03, 5.42729481e-01, 6.90482557e+02],
        [8.97907155e+00, 1.30051581e+02, 6.79652537e-01, 1.06226560e+04,
         6.62587378e+02, 1.47691095e-01, 1.87898628e+02]]))

The first value (12698.78) is the Chi-square value, followed by the p-value ~ 0, then comes the degrees of freedom (18), and lastly it outputs the expected frequencies as an array. 
Since all of the expected frequencies are greater than 5, the Chi-square test results can be trusted.
We can reject the null hypothesis as the p-value is less than 0.05. 
Thus, the results indicate that there is a correlation/some sort of relationship between payment type and borough. For sure we do know that these two variables are not independent of each other.

We could relate this result to the "industrialization" of boroughs. Focusing on Manhattan, Brooklyn, EWR and Queens we can see that the most common way of payment is with credit card. Fos Staten Island there's no big difference and in Bronx most of the people pay with cash.

In [ ]:
###Grafico per visualizzare il numero di pagamenti per ogni mezzo###

#contingency_table = pd.crosstab(datazone['payment_type'], datazone['Borough'])
#Assigns the frequency values
#creditcard = contingency_table.iloc[0][0:6].values
#cash = contingency_table.iloc[1][0:6].values
#Plots the bar chart
#fig = plt.figure(figsize=(10, 5))
#sns.set(font_scale=1.5)
#categories = ["Bronx","Brooklyn","Manhattan","Queens","Staten Island","Unknown"]
#p1 = plt.bar(categories, creditcard, color='#d62728')
#p2 = plt.bar(categories, cash, bottom=creditcard)
#plt.legend((p2[0], p1[0]), ('Cash', 'Credit Card'))
#plt.xlabel('Borough')
#plt.ylabel('Count')
#plt.show()